flips/20211129-cadence-mutability-restrictions.md

```diff
@@ -0,0 +1,417 @@
+mutated in all contexts, but not written, while `pub(set) var` created
+a field that could be read, written and mutated in all contexts. After this change,
+it would not be possible to describe a field that can be read and mutated but not set, 
+as `pub let` would only allow reads in arbitrary contexts. 
+
+### Alternatives Considered
+
+One possible approach would be to add mutability modifiers to field or variable declarations,
+like the `readonly` or `mut` tags found in languages like Rust, C# and TypeScript. This would make
+all fields either mutable or immutable by default, requiring users to supply the appropriate tag
+to make their behavior be whichever is not the default. 
+
+This approach has the unfortunate side effect of increasing the surface area of the language, while
+not also adding immediate benefit; the `readonly` and `mut` distinction may be redundant when
+`let` and `var` already exist in the language. To justify this addition, we would need to identify
+a use case for allowing field to be written to but not mutated (`readonly var`), or for restricting 
+the mutation of fields even within their enclosing type. 
+
+Another approach would be to make `let` a true constant declaration, and forbid writing or mutating it in any context,
+while allowing `var` to be mutated in all the contexts it can be written. This, however, is likely too restrictive;
+a complex initialization for a dictionary or array value may benefit from the ability to iteratively mutate its contents,
+up to the point at which its value is finalized, and can thus be exported. As such we would like to still allow users
+to perform mutations in an limited context.
+
+A third approach would be to ban contract-level public field, with the idea that users cannot accidentally shoot 
+themselves in the foot by exporting a `pub let` field from their contract and having it be externally mutated 
+if `pub let` fields cannot exist on contracts in the first place. However, this has a number of downsides. The
+obvious one is that all data that one might wish to expose to be read from a contract would need an explicit getter
+method. The second, less obvious but also more concerning downside, is that it would be necessary to 
+ban `pub let` fields on structs and resources as well. Consider the case where a user exports a getter method
+to read a struct or resource used by their contract. Any `pub let` fields on this struct or resource 
+would also be mutable, and thus are subject to the same risks they would be if they were exported
+directly from the contract. 
+
+Consider:
+
+```
+pub contract C {
+    pub struct Foo {
+        pub let arr : [Int]
+        init() {
+            self.arr = [3]
+        }
+    }
+    
+    priv let foo : Foo
+
+    init() {
+        self.foo = Foo()
+    }
+
+    pub fun getFoo(): Foo {
+        return self.foo
+    }
+}
+
+pub fun main() {
+    let a = C.getFoo()
+    a.arr.append(0) // a.arr is now [3, 0]
```

Comment:
Is this what is proposed here? Currently, this should not be expected, as the field's value is copied

Reply:
When I run this code as a contract/transaction pair on the playground, the output is `[3,0]`. Is that not expected?

```
pub contract C {
    pub struct Foo {
        pub let arr : [Int]
        init() {
            self.arr = [3]
        }
    }
    
    priv let foo : Foo
    init() {
        self.foo = Foo()
    }
    pub fun getFoo(): Foo {
        return self.foo
    }
}
```
and 
```
import C from 0x01

transaction {

  prepare(acct: AuthAccount) {}

  execute {
    let a = C.getFoo()
    a.arr.append(0) // a.arr is now [3, 0]
    log(a.arr)
  }
}
```

Reply:
Ah, my bad, I misread the example as the array being returned from the getter and being modified!